### PR TITLE
Certificate validation callback

### PR DESCRIPTION
**Pull request by carlosmn**
The first commit is just so I can run it on my machines, the others will get squashed.

I'm not sure how deeply to specify types for the certs. ~~~Maybe it'd be fine to expose the cert like in C where the same class may contain MD5 or SHA1 hashes.~~~ It looks like I forgot that it's theoretically possible to have both hostkey hashes at once, so it should be just the one.

I don't quite expect this to actually work yet, but the types and the structs should at least be correct.

- [x] Test an X.509 certificate
- [x] Test a ssh certificate

### CODE REVIEW

**Comment by carlosmn** [comment]
Should be ready for review. I don't know if it's worth having a different test for accepting vs rejecting. I'm mostly concerned about exposing the data to the user here.

File: LibGit2Sharp/CertificateSsh.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        {
+
+            HasMD5  = cert.type.HasFlag(GitCertificateSshType.MD5);
+            HasSHA1 = cert.type.HasFlag(GitCertificateSshType.SHA1);
+
+            HashMD5 = new byte[16];
+            cert.HashMD5.CopyTo(HashMD5, 0);
+
+            HashSHA1 = new byte[20];
+            cert.HashSHA1.CopyTo(HashSHA1, 0);
```
Comment (carlosmn) [inline LibGit2Sharp/CertificateSsh.cs]: Since we've brought these arrays into C# already, I wonder if it'd be fine to assing `HashMD5 = cert.HashMD5` and the same for SHA1.

**Comment by nulltoken** [comment]
pushed up some minor fixups.

File: LibGit2Sharp.Tests/CloneFixture.cs [hunk 2]
```diff
@@ -237,6 +237,44 @@ public void CanCloneFromBBWithCredentials(string url, string user, string pass,
...
         }
 
+        [Theory]
+        [InlineData("https://github.com/libgit2/TestGitRepository.git", "github.com", typeof(CertificateX509))]
+        [InlineData("git@github.com/libgit2/TestGitRepository.git", "github.com", typeof(CertificateSsh))]
+        public void CanInspectCertificateOnClone(string url, string hostname, Type certType)
+        {
+            var scd = BuildSelfCleaningDirectory();
+
+            Assert.Throws<UserCancelledException>(() => {
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/CloneFixture.cs]: I'm not sold on this. I find this code construct very difficult to read.

This should typically be a dynamically skipped test. But as this is not a thing any longer in xUnit 2.0, we should come with a  more straightforward way to express this.

/cc @dahlbyk @Therzok Any idea?

File: LibGit2Sharp/LibGit2Sharp.csproj [hunk 3]
```diff
@@ -23,7 +23,6 @@
     <DefineConstants>TRACE;DEBUG;NET40</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
Comment (nulltoken) [inline LibGit2Sharp/LibGit2Sharp.csproj]: Any chance to move those changes into a separate commit?

File: LibGit2Sharp.Tests/CloneFixture.cs (hunk 2 above)
Comment (nulltoken) [inline LibGit2Sharp.Tests/CloneFixture.cs]: One of my main issue with this Skipping feature no longer being doable with xUnit 2.0 is that a passing test may now "mean" two completely different thing.

 - Yeah, everything worked perfectly.
 - Sorry, dude, I bailed out.

And *that* somehow bothers me.

File: LibGit2Sharp/LibGit2Sharp.csproj (hunk 3 above)
Comment (Therzok) [inline LibGit2Sharp/LibGit2Sharp.csproj]: Why change this? Leave it as it was.

Comment (Therzok) [inline LibGit2Sharp/LibGit2Sharp.csproj]: This way we ensure build breaks on missing XML doc and other warnings (that are usually not false-positives)

File: LibGit2Sharp.Tests/CloneFixture.cs (hunk 2 above)
Comment (Therzok) [inline LibGit2Sharp.Tests/CloneFixture.cs]: We still have the answer to [your question](http://stackoverflow.com/a/27832300) here.

File: LibGit2Sharp/LibGit2Sharp.csproj (hunk 3 above)
Comment (carlosmn) [inline LibGit2Sharp/LibGit2Sharp.csproj]: This is just what MD/XS does to the config file, I'd just been committing everything.

Comment (Therzok) [inline LibGit2Sharp/LibGit2Sharp.csproj]: 6.0 should fix this issue since we're changing the way the project model is serialized.

File: LibGit2Sharp.Tests/CloneFixture.cs (hunk 2 above)
Comment (carlosmn) [inline LibGit2Sharp.Tests/CloneFixture.cs]: I don't like that bit either, but as I had to use the xunit2 port, this was the best I could come up with to test that the ssh "cert" is in order without failing the test when built without ssh support.

Comment (nulltoken) [inline LibGit2Sharp.Tests/CloneFixture.cs]: I've updated https://github.com/xunit/samples.xunit/issues/8 to express the kind of use we'd make of such a dynamic ignore.

@bradwilson's answer left me under the impression that this *might* be doable with xUnit 2.0. I may have misunderstood it, though.

**Comment by bradwilson** [comment]
The SkippableFact example is now available. https://github.com/xunit/samples.xunit/tree/master/DynamicSkipExample

**Comment by Therzok** [comment]
:green_heart:

**Comment by nulltoken** [comment]
@bradwilson AWE. SOME. :bangbang:

**Comment by nulltoken** [comment]
@carlosmn Now that @bradwilson has sprinkled some of his amazing magic, could you please update this to rely on the `[SkippableFact]` attribute and maybe update #895 with the xUnit 2.0 compatible implementation?

**Comment by carlosmn** [comment]
@nulltoken updated to use `SkippableFact`. This happily works with xuni1 as well, so this shouldn't depend on the xunit2 branch to go in.

**Comment by nulltoken** [comment]
Pushed up a counter proposal. This actually fails. Any idea?

**Comment by nulltoken** [comment]
@carlosmn Can you please check the `GitCertificateType` xml doc of the additions?

File: LibGit2Sharp.Tests/CloneFixture.cs [hunk 4]
```diff
@@ -237,6 +237,54 @@ public void CanCloneFromBBWithCredentials(string url, string user, string pass,
...
+                        // we get a string with the different fields instead of a structure, so...
+                        Assert.True(x509.Subject.Contains("CN=github.com,"));
+                        return true;
+                    }
+
+                    if (certType == typeof(CertificateSsh))
+                    {
+                        var hostkey = (CertificateSsh)cert;
+                        Assert.True(hostkey.HasMD5);
+                        Assert.Equal("1627aca576282d36631b564debdfa648", BitConverter.ToString(hostkey.HashMD5));
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/CloneFixture.cs]: Could you please add a comment explaining how to discover the new hash would the hostkey change?

File: LibGit2Sharp/CertificateSsh.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        /// <summary>
+        /// The SHA1 hash of the host. Meaningful if <see cref="HasSHA1"/> is true
+        /// </summary>
+        public readonly byte[] HashSHA1;
+
+        /// <summary>
+        /// True if we have the MD5 hostkey hash from the server
+        /// </summary>
+        public readonly bool HasMD5;
```
Comment (nulltoken) [inline LibGit2Sharp/CertificateSsh.cs]: How many kinds of Hash can we get? Rather then exposing named properties, can't we this into an enum?

File: LibGit2Sharp/CertificateSsh.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        /// <summary>
+        /// For mocking purposes
+        /// </summary>
+        protected CertificateSsh()
+        { }
+
+        /// <summary>
+        /// The MD5 hash of the host. Meaningful if <see cref="HasMD5"/> is true
+        /// </summary>
+        public readonly byte[] HashMD5;
```
Comment (nulltoken) [inline LibGit2Sharp/CertificateSsh.cs]: How dumb would is it to convert this into a readable string?

File: LibGit2Sharp/CertificateSsh.cs (hunk 5 above)
Comment (nulltoken) [inline LibGit2Sharp/CertificateSsh.cs]: And only expose a `Hash` property?

Comment (carlosmn) [inline LibGit2Sharp/CertificateSsh.cs]: We should be able to get both. My ssh and libssh2 only seem to give back MD5 for the servers I connect to, but it claims it could be both.

File: LibGit2Sharp/CertificateSsh.cs (hunk 6 above)
Comment (carlosmn) [inline LibGit2Sharp/CertificateSsh.cs]: Which one, though? There's at least three ways to represent this string. OpenSSH does it lowercase with colons between each byte, I did it without the colons in libgit2 testing to abuse the oid functions, and `BitConverter` does uppercase with dashes for each byte.

Comment (nulltoken) [inline LibGit2Sharp/CertificateSsh.cs]: @Therzok Thoughts?

Comment (Therzok) [inline LibGit2Sharp/CertificateSsh.cs]: Why not let the user decide?

Comment (nulltoken) [inline LibGit2Sharp/CertificateSsh.cs]: How so?

Comment (Therzok) [inline LibGit2Sharp/CertificateSsh.cs]: Leave it as byte array, I mean. And provide convenience functions for transformation?

Comment (carlosmn) [inline LibGit2Sharp/CertificateSsh.cs]: That was my rationale. It's a bunch of bytes, let's present it that way. If you want to ask the user, it's up to you to create the string representation.

Comment (carlosmn) [inline LibGit2Sharp/CertificateSsh.cs]: You might also wish to note that `System.Security.Cryptography.HashAlgorithm` (from which C#'s hash algo classes inherit) uses a `byte[]` to represent the resulting hash.

Comment (ethomson) [inline LibGit2Sharp/CertificateSsh.cs]: > Why not let the user decide?

:+1:

**Comment by nulltoken** [comment]
:gem:

**Comment by michaKFromParis** [comment]
Great! Thanks for merging.

**Comment by nulltoken** [comment]
@michaKFromParis I'm going to publish in the following hours a new pre-release version of the NuGet package containing this feature.

**Comment by nulltoken** [comment]
Published as NuGet pre-release package `LibGit2Sharp.0.22.0-pre20150716071016`
